packages/core/saved-objects/core-saved-objects-server/docs/model_versions.md

```diff
@@ -0,0 +1,612 @@
+const change: SavedObjectsModelMappingsAdditionChange = {
+  type: 'mappings_addition',
+  addedMappings: {
+    newField: { type: 'text' },
+    existingNestedField: {
+      properties: {
+        newNestedProp: { type: 'keyword' },
+      },
+    },
+  },
+};
+```
+
+**note:** *When adding mappings, the root `type.mappings` must also be updated accordingly (as it was done previously).*
+
+#### - mappings_deprecation
+
+Used to flag mappings as no longer being used and ready to be removed.
+
+*Usage example:*
+
+```ts
+let change: SavedObjectsModelMappingsDeprecationChange = {
+  type: 'mappings_deprecation',
+  deprecatedMappings: ['someDeprecatedField', 'someNested.deprecatedField'],
+};
+```
+
+**note:** *It is currently not possible to remove fields from an existing index's mapping (without reindexing into another index),
+          so the mappings flagged with this change type won't be deleted for now, but this should still be used to allow
+          our system to clean the mappings once upstream (ES) unblock us.*
+
+#### - data_backfill
+
+Used to populate fields (indexed or not) added in the same version
+
+*Usage example:*
+
+```ts
+let change: SavedObjectsModelDataBackfillChange = {
+  type: 'data_backfill',
+  transform: (document) => {
+    document.attributes.someAddedField = 'defaultValue';
+    return { document };
+  },
+};
+```
+
+**note:** *Even if no check is performed to ensure it, this type of model change should only be used to 
+           backfill newly introduced fields.*
+
+### schemas
+
+[link to the TS doc for `schemas`](https://github.com/elastic/kibana/blob/9b330e493216e8dde3166451e4714966f63f5ab7/packages/core/saved-objects/core-saved-objects-server/src/model_version/schemas.ts#L11-L16)
+
+The schemas associated with this version. Schemas are used to validate or convert SO documents at various
+stages of their lifecycle.
+
+The currently available schemas are:
+
+#### forwardCompatibility
+
+This is a new concept introduced by model versions. This schema is used for inter-version compatibility.
+
+When retrieving a savedObject document from an index, if the version of the document is higher than the latest version
+known of the Kibana instance, the document will go through the `forwardCompatibility` schema of the associated model version.
+
+**Important:** These conversion mechanism shouldn't assert the data itself, and only strip unknown fields to convert the document to 
+the **shape** of the document at the given version.
+
+Basically, this schema should keep all the known fields of a given version, and remove all the unknown fields, without throwing.
+
+Forward compatibility schema can be implemented in two different ways.
+
+1. Using `config-schema`
+
+*Example of schema for a version having two fields: someField and anotherField*
+
+```ts
+const versionSchema = schema.object(
+  {
+    someField: schema.maybe(schema.string()),
+    anotherField: schema.maybe(schema.string()),
+  },
+  { unknowns: 'ignore' }
```

Comment:
I wish we could do this by default or, at the very least, warn when the 'forwardCompatibility' versionSchema doesn't declare `unknowns: 'ignore'`

Reply:
100%, I would very much have a safer, less developer-error-prone way to define this specific schema. I couldn't find any obvious quick win around that aspect though. Ideally, mid/long term, we'd use an higher level description that automatically generates the config-schema validation...